ktlint removes first U+FEFF seen anywhere in its input, even if it's in a literal

https://github.com/pinterest/ktlint/pull/630 made ktlint temporarily remove byte order marks, and later put them back at the start of the file.  The way the removal in particular works seems to produce corner cases whereby ktlint can accidentally remove U+FEFF where it shouldn't, and not put them back.

When this U+FEFF is inside a char or string literal, this is unsound: the contents of the literal are changed by formatting.  Worse, if it's inside a char literal, the formatting process makes the literal empty and therefore _syntactically_ invalid.

While this problem relates to a character known to be problematic and to-be-avoided in UTF8 input, and can be worked around by always using escape sequences, it can be triggered for example (and this is how I encountered the bug!) by code generators that emit lightly-escaped literal UTF-8 data into string literals.

This behaviour seems to originate from [here](https://github.com/pinterest/ktlint/blob/76ecdfc9d18c49ff15fb9394b069493e8b5660c1/ktlint-rule-engine/src/main/kotlin/com/pinterest/ktlint/rule/engine/internal/RuleExecutionContext.kt#L262-L266), but as an outsider I suspect I may be missing some context as to why the problem of BOMs appearing in input code was solved this way.

## Expected Behavior

ktlint should only disturb U+FEFF at the start of an input script, as (to the best of my knowledge) this is the only place where it could possibly be a byte-order mark.  In no circumstances should it remove U+FEFF in char and string literals.

## Observed Behavior

ktlint can remove, and fail to replace, the first U+FEFF in the input script, even if it isn't at the start of the input.

## Steps to Reproduce

1. Create a KotlinScript file with both a code style violation and a `U+FEFF` character (not an escape sequence) inside something (such as a string or char literal) that shouldn't be changed by ktlint.  The example I have is:

```kotlin
println ('﻿')
```
(where the space before `(` is a style violation, and there is a `U+FEFF` character between the parentheses; IDEA shows this as `[ZWNBSP]`).

This produces the following hex dump:

```shell
$ xxd bug.kts
00000000: 7072 696e 746c 6e20 2827 efbb bf27 290a  println ('...').
```

Note the `efbb bf` in the hex dump, which is the UTF-8 encoding of `U+FEFF`.

In IDEA, this file is syntactically valid and executable, with the expected result.

2. Run ktlint in format mode on the file:

```shell
$ ktlint -F bug.kts
```

3. Observe (eg. in a hex editor or in IDEA) that the code has been formatted to

```kotlin
println('')
```

but the `U+FEFF` character has disappeared:

```shell
$ xxd bug.kts
00000000: 7072 696e 746c 6e28 2727 290a            println('').
```

This file is now syntactically invalid and can't be run from within IDEA:

```
bug.kts:1:9: error: empty character literal.
println('')
        ^
```

Note that if the file _starts_ with a BOM, ktlint works as expected (fixes the formatting but preserves all `U+FEFF`):

```shell
$  xxd bug.kts
00000000: efbb bf70 7269 6e74 6c6e 2028 27ef bbbf  ...println ('...
00000010: 2729 0a                                  ').
$ ktlint -F bug.kts
$ xxd bug.kts
00000000: efbb bf70 7269 6e74 6c6e 2827 efbb bf27  ...println('...'
00000010: 290a                                     ).
```

## Your Environment

```shell
$ ktlint --version
ktlint version 1.8.0
```

Observed in CLI usage with no `.editorconfig` on macOS Sequoia 15.7.3, system java is openjdk 11.0.24 2024-07-16 LTS.  Originally seen in a custom integration of a `KtLintRuleEngine` against ktlint 1.4.0.

--- a/ktlint-rule-engine/src/main/kotlin/com/pinterest/ktlint/rule/engine/internal/RuleExecutionContext.kt
+++ b/ktlint-rule-engine/src/main/kotlin/com/pinterest/ktlint/rule/engine/internal/RuleExecutionContext.kt
@@ -260,7 +260,13 @@ internal class RuleExecutionContext private constructor(
             text
                 .replace("\r\n", "\n")
                 .replace("\r", "\n")
-                .replaceFirst(UTF8_BOM, "")
+                .let {
+                    if (it.startsWith(UTF8_BOM)) {
+                        it.replaceFirst(UTF8_BOM, "")
+                    } else {
+                        it
+                    }
+                }
 
         private fun PsiElement.findErrorElement(): PsiErrorElement? {
             if (this is PsiErrorElement) {

--- a/ktlint-rule-engine/src/test/kotlin/com/pinterest/ktlint/rule/engine/api/KtLintTest.kt
+++ b/ktlint-rule-engine/src/test/kotlin/com/pinterest/ktlint/rule/engine/api/KtLintTest.kt
@@ -27,6 +27,7 @@ import com.pinterest.ktlint.rule.engine.core.api.editorconfig.EditorConfig
 import com.pinterest.ktlint.rule.engine.core.api.ifAutocorrectAllowed
 import com.pinterest.ktlint.rule.engine.core.api.isRoot20
 import com.pinterest.ktlint.rule.engine.core.api.replaceTextWith
+import com.pinterest.ktlint.ruleset.standard.rules.IndentationRule
 import org.assertj.core.api.Assertions.assertThat
 import org.ec4j.core.model.PropertyType
 import org.jetbrains.kotlin.com.intellij.lang.ASTNode
@@ -226,19 +227,22 @@ class KtLintTest {
     }
 
     @Test
-    fun testFormatUnicodeBom() {
+    fun `Given a file that starts with the UTF8 BOM character then ensure that the formatted file starts with that character as well`() {
         val code =
-            getResourceAsText("spec/format-unicode-bom.kt.spec")
+            getResourceAsText("spec/format-unicode-bom-at-start-of-file--before-ktlint-format.kt")
+                // Standardize code to use LF as line separator regardless of OS
+                .replace("\r\n", "\n")
+                .replace("\r", "\n")
+        val formattedCode =
+            getResourceAsText("spec/format-unicode-bom-at-start-of-file--after-ktlint-format.kt")
                 // Standardize code to use LF as line separator regardless of OS
                 .replace("\r\n", "\n")
                 .replace("\r", "\n")
 
         val actual =
             KtLintRuleEngine(
-                ruleProviders =
-                    setOf(
-                        RuleV2InstanceProvider { DummyRule() },
-                    ),
+                // We need a rule that actually modifies the file to test whether UTF BOM character is handled correctly
+                ruleProviders = setOf(RuleV2InstanceProvider { IndentationRule() }),
                 editorConfigOverride =
                     EditorConfigOverride.from(
                         // The code sample use LF as line separator, so ensure that formatted code uses that as well, as otherwise the test
@@ -247,7 +251,35 @@ class KtLintTest {
                     ),
             ).format(Code.fromSnippet(code)) { _ -> AutocorrectDecision.ALLOW_AUTOCORRECT }
 
-        assertThat(actual).isEqualTo(code)
+        assertThat(actual).isEqualTo(formattedCode)
+    }
+
+    @Test
+    fun `Issue 3220 - Given some code that contains the UTF8 BOM character somewhere but not a the start then ensure this character is not removed`() {
+        val code =
+            getResourceAsText("spec/do-not-format-unicode-bom-when-not-at-start-of-file--before-ktlint-format.kt")
+                // Standardize code to use LF as line separator regardless of OS
+                .replace("\r\n", "\n")
+                .replace("\r", "\n")
+        val formattedCode =
+            getResourceAsText("spec/do-not-format-unicode-bom-when-not-at-start-of-file--after-ktlint-format.kt")
+                // Standardize code to use LF as line separator regardless of OS
+                .replace("\r\n", "\n")
+                .replace("\r", "\n")
+
+        val actual =
+            KtLintRuleEngine(
+                // We need a rule that actually modifies the file to test whether UTF BOM character is handled correctly
+                ruleProviders = setOf(RuleV2InstanceProvider { IndentationRule() }),
+                editorConfigOverride =
+                    EditorConfigOverride.from(
+                        // The code sample use LF as line separator, so ensure that formatted code uses that as well, as otherwise the test
+                        // breaks on Windows OS
+                        END_OF_LINE_PROPERTY to PropertyType.EndOfLineValue.lf,
+                    ),
+            ).format(Code.fromSnippet(code)) { _ -> AutocorrectDecision.ALLOW_AUTOCORRECT }
+
+        assertThat(actual).isEqualTo(formattedCode)
     }
 
     @Nested

--- a/ktlint-rule-engine/src/test/resources/spec/.editorconfig
+++ b/ktlint-rule-engine/src/test/resources/spec/.editorconfig
@@ -1,0 +1,2 @@
+[*.{kt,kts}]
+ktlint_standard = disabled

--- a/ktlint-rule-engine/src/test/resources/spec/do-not-format-unicode-bom-when-not-at-start-of-file--after-ktlint-format.kt
+++ b/ktlint-rule-engine/src/test/resources/spec/do-not-format-unicode-bom-when-not-at-start-of-file--after-ktlint-format.kt
@@ -1,0 +1,2 @@
+// Although probably not visible in your editor, this file contains at UTF BOM --> ﻿ <--
+// The indent before this comment needs to be removed by ktlint format

--- a/ktlint-rule-engine/src/test/resources/spec/do-not-format-unicode-bom-when-not-at-start-of-file--before-ktlint-format.kt
+++ b/ktlint-rule-engine/src/test/resources/spec/do-not-format-unicode-bom-when-not-at-start-of-file--before-ktlint-format.kt
@@ -1,0 +1,2 @@
+// Although probably not visible in your editor, this file contains at UTF BOM --> ﻿ <--
+    // The indent before this comment needs to be removed by ktlint format

--- a/ktlint-rule-engine/src/test/resources/spec/format-unicode-bom-at-start-of-file--after-ktlint-format.kt
+++ b/ktlint-rule-engine/src/test/resources/spec/format-unicode-bom-at-start-of-file--after-ktlint-format.kt
@@ -1,1 +1,2 @@
 ﻿// Although probably not visible in your editor, this file starts with a UTF8 BOM unicode character
+// The indent before this comment needs to be removed by ktlint format

--- a/ktlint-rule-engine/src/test/resources/spec/format-unicode-bom-at-start-of-file--before-ktlint-format.kt
+++ b/ktlint-rule-engine/src/test/resources/spec/format-unicode-bom-at-start-of-file--before-ktlint-format.kt
@@ -1,0 +1,2 @@
+﻿// Although probably not visible in your editor, this file starts with a UTF8 BOM unicode character
+    // The indent before this comment needs to be removed by ktlint format

--- a/ktlint-rule-engine/src/test/resources/spec/readme.md
+++ b/ktlint-rule-engine/src/test/resources/spec/readme.md
@@ -1,0 +1,31 @@
+The spec files in this directory contain the UTF BOM character \uFEFF
+
+The files have been created with Zsh terminal commands below:
+```shell
+echo -e "\uFEFF// Although probably not visible in your editor, this file starts with a UTF8 BOM unicode character\n    // The indent before this comment needs to be removed by ktlint format" > format-unicode-bom-at-start-of-file-before-format.kt.spec
+echo -e "\uFEFF// Although probably not visible in your editor, this file starts with a UTF8 BOM unicode character\n// The indent before this comment needs to be removed by ktlint format" > format-unicode-bom-at-start-
+of-file-after-format.kt.spec
+
+echo -e "// Although probably not visible in your editor, this file contains at UTF BOM --> \uFEFF <--\n    // The indent before this comment needs to be removed by ktlint format" > do-not-format-unicode-bom-when-not-at-start-of-file-before-ktlint-format.kt.spec
+echo -e "// Although probably not visible in your editor, this file contains at UTF BOM --> \uFEFF <--\n// The indent before this comment needs to be removed by ktlint format" > do-not-format-unicode-bom-when-not-at-st
+art-of-file--after-ktlint-format.kt.spec
+```
+
+With the `xxd` command the files can be inspected:
+```shell
+xxd do-not-format-unicode-bom-when-not-at-start-of-file--before-ktlint-format.kt
+```
+results in output like:
+```terminaloutput
+00000000: 2f2f 2041 6c74 686f 7567 6820 7072 6f62  // Although prob
+00000010: 6162 6c79 206e 6f74 2076 6973 6962 6c65  ably not visible
+00000020: 2069 6e20 796f 7572 2065 6469 746f 722c   in your editor,
+00000030: 2074 6869 7320 6669 6c65 2063 6f6e 7461   this file conta
+00000040: 696e 7320 6174 2055 5446 2042 4f4d 202d  ins at UTF BOM -
+00000050: 2d3e 20ef bbbf 203c 2d2d 0a20 2020 202f  -> ... <--.    /    <<-- The "ef bb" corresponds with the \uFEFF character
+00000060: 2f20 5468 6520 696e 6465 6e74 2062 6566  / The indent bef
+00000070: 6f72 6520 7468 6973 2063 6f6d 6d65 6e74  ore this comment
+00000080: 206e 6565 6473 2074 6f20 6265 2072 656d   needs to be rem
+00000090: 6f76 6564 2062 7920 6b74 6c69 6e74 2066  oved by ktlint f
+000000a0: 6f72 6d61 740a                           ormat.
+```

